modules/database/src/std/filters/ts.c

```diff
@@ -56,34 +160,207 @@ static db_field_log* filter(void* pvt, dbChannel *chan, db_field_log *pfl) {
     return pfl;
 }
 
-static void channelRegisterPre(dbChannel *chan, void *pvt,
-                               chPostEventFunc **cb_out, void **arg_out, db_field_log *probe)
-{
+static db_field_log *replace_fl_value(tsPrivate const *pvt,
+                                      db_field_log *pfl,
+                                      void (*func)(tsPrivate const *,
+                                                   db_field_log *)) {
+    /* Get rid of the old value */
+    if (pfl->type == dbfl_type_ref && pfl->u.r.dtor) {
+        pfl->u.r.dtor(pfl);
+    }
+    pfl->no_elements = 1;
+    pfl->type = dbfl_type_val;
+
+    func(pvt, pfl);
+    return pfl;
+}
+
+static void ts_to_array(tsPrivate const *settings,
+                        epicsTimeStamp const *ts,
+                        epicsUInt32 arr[2]) {
+    arr[0] = ts->secPastEpoch;
+    arr[1] = ts->nsec;
+    if (settings->epoch == tsEpochUnix) {
+        /* Cannot use epicsTimeToWhatever because Whatever uses signed ints */
+        arr[0] += POSIX_TIME_AT_EPICS_EPOCH;
+    }
+}
+
+static void ts_seconds(tsPrivate const *settings, db_field_log *pfl) {
+    epicsUInt32 arr[2];
+    ts_to_array(settings, &pfl->time, arr);
+    pfl->field_type = DBF_ULONG;
+    pfl->field_size = sizeof(epicsUInt32);
+    pfl->u.v.field.dbf_ulong = arr[0];
+}
+
+static void ts_nanos(tsPrivate const *settings, db_field_log *pfl) {
+    epicsUInt32 arr[2];
+    ts_to_array(settings, &pfl->time, arr);
+    pfl->field_type = DBF_ULONG;
+    pfl->field_size = sizeof(epicsUInt32);
+    pfl->u.v.field.dbf_ulong = arr[1];
+}
+
+static void ts_double(tsPrivate const *settings, db_field_log *pfl) {
+    epicsUInt32 arr[2];
+    ts_to_array(settings, &pfl->time, arr);
+    pfl->field_type = DBF_DOUBLE;
+    pfl->field_size = sizeof(epicsFloat64);
+    pfl->u.v.field.dbf_double = arr[0] + arr[1] * 1e-9;
+}
+
+static void ts_array(tsPrivate const *settings, db_field_log *pfl) {
+    pfl->field_type = DBF_ULONG;
+    pfl->field_size = sizeof(epicsUInt32);
+    pfl->no_elements = 2;
+    pfl->type = dbfl_type_ref;
+    pfl->u.r.pvt = NULL;
+    pfl->u.r.field = allocTsArray();
```

Comment:
`freeListCalloc()` can return NULL, which needs to be handled in some way.  (eg. return 0 elements?)

Reply:
Good point, I'm too used to OS where allocations can't fail ...